### PR TITLE
Support sharded safetensors in TF

### PR DESCRIPTION
**Pull request by Rocketknight1**
Right now our TF safetensors loading doesn't support sharded checkpoints, which is a problem as more and more big models move to safetensors weights only! This is currently blocking @a8nova's PR at #26870.

As sharded safetensors **saving** for TF was also missing, this PR adds that as well, and expands the tests to cover both.

TODO:

- [x] Check we prioritize safetensors vs. TF weights correctly when both are present (and make sure behaviour matches for sharded vs. unsharded)
- [x] Add tests for sharded safetensors loading from both PT and TF format
- [x] Add test for saving sharded safetensors

### CODE REVIEW

**Comment by Rocketknight1** [comment]
This should be ready for review! cc @ArthurZucker @a8nova

File: tests/test_modeling_tf_utils.py
Comment (LysandreJik) [inline tests/test_modeling_tf_utils.py:368]: `safetensors` is now a base dependency so maybe we should eventually just remove all of these

**Review by LysandreJik — APPROVED** [review]
Looks good! When would you have in mind an eventual switch to safetensors serialization by default?

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_pytorch_utils.py:391]: It is, unfortunately! The reason is that this function is used both to load shards and non-sharded checkpoints. When it's loading a non-sharded checkpoint, we want to log missing keys immediately. When it's loading a shard, there will always be lots of "missing" keys, but we don't want to log those - instead, we only want to log keys that are missing from **every** shard, which we will only know after all shards have been loaded. This is handled in the sharded loading function.

File: src/transformers/modeling_tf_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:2925]: Added!

File: tests/test_modeling_tf_utils.py
Comment (Rocketknight1) [reply tests/test_modeling_tf_utils.py:368]: Makes sense - do you want me to just do it in this PR?

**Comment by Rocketknight1** [comment]
@LysandreJik I think now that we have proper support we can switch to safetensors by default immediately, either in this PR or in a follow-up.

**Comment by LysandreJik** [comment]
I'd wait for a few weeks just to ensure we don't have reports of failure and switch for the next version. WDYT?

**Comment by Rocketknight1** [comment]
Sounds good to me!

**Comment by Rocketknight1** [comment]
@LysandreJik is there anything else to be resolved before I merge this? (Except for the failing test, but that's not specific to this PR)

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (amyeroberts) [inline src/transformers/modeling_tf_pytorch_utils.py:476]: AFAICT, all these checks are the same as the ones above. Can we abstract these out to e.g. `validate_keys(unmatched_keys, missing_keys, mismatched_keys) and call that in both of the functions?

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (amyeroberts) [inline src/transformers/modeling_tf_pytorch_utils.py:475]: My understanding is that we want to have skip_logger_warning=True when calling `load_pytorch_state_dict_in_tf2_model` here, but I don't see why we're enabling skipping here? Silencing the logging warnings should really be a hidden functionality (maybe with a param `_skip_logger_warnings` and not something people calling either function use

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this and enabling sharded support! 

Just a few small comments / questions

File: tests/test_modeling_tf_utils.py
Comment (Rocketknight1) [reply tests/test_modeling_tf_utils.py:536]: Re-added!

File: tests/test_modeling_tf_utils.py
Comment (Rocketknight1) [reply tests/test_modeling_tf_utils.py]: Good catch - that was copied from the h5 test, and wouldn't work for safetensors - we just got lucky that it wasn't called in the tests anyway.

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_pytorch_utils.py]: Done!

**Comment by Rocketknight1** [comment]
All comments addressed @amyeroberts! I think we should be ready now.

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (amyeroberts) [inline src/transformers/modeling_tf_pytorch_utils.py:405]: nit - definition should go above the lines of code where it's used i.e. before load_pytorch_state_dict_in_tf2_model

File: src/transformers/modeling_tf_utils.py
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:650]: Should we have the default shard size match the one in load_tf_weights?

File: src/transformers/modeling_tf_utils.py
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:2997]: nit - comment isn't really necessary here - the function name is pretty explicit 

```suggestion
```

File: src/transformers/modeling_tf_utils.py
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:2925]: Looks like we might want a constant e.g. `SUPPORTED_SAFE_FORMATS = ["pt", "tf", "flax", "mlx"]` so we don't have to update this in several locations here and for PT (for another PR)

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this support!

File: src/transformers/modeling_tf_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:2925]: Agreed!

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_pytorch_utils.py]: Done!

File: src/transformers/modeling_tf_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:650]: I got here by following the torch code, which also has the same issue! Specifically, `save_pretrained()` has a default size of 5GB, but the actual checkpoint sharding methods have a default size of 10GB. In general, though, the value passed from `save_pretrained()` will override those values.

It's a very minor detail either way, since I think both 5GB and 10GB shards work fine! We could consider standardizing everything at some point, but I don't think it's a high priority.

**Comment by Rocketknight1** [comment]
This looks ready to go, cc @a8nova! We have a branch cut + release planned on Monday, though, and since this touches a lot of core code I don't want to merge it right before a release. Instead, I suggest merging it right after the branch cut, and then we can finalize and merge the PRs that are blocked by it: TF-IDEFICS, TF-Gemma and possibly the Mistral/Mixtral PRs if @ariG23498 can have one of them ready by then (no stress, obviously!)

Then we could launch all the new Keras models together in the following release and do a section in the release notes about them, crediting @a8nova and @ariG23498?

**Comment by Rocketknight1** [comment]
cc @a8nova and @ariG23498, this has now been merged. If you rebase your PRs, that should resolve any issues with sharded safetensors loading!

**Pull request merged**
